### ISSUE TITLE
improve nn decision of layout ai

### ISSUE THREAD

**Issue by Shinichi0713** [Shinichi0713/RL-study#155]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by Shinichi0713** [comment]
seeing the result, the deiversity of agent decision is lost.

**Comment by Shinichi0713** [comment]
there, the selected action is biased.  

![Image](https://github.com/user-attachments/assets/8b2b7601-b983-42a5-8059-3aeedf48de8f)

**Comment by Shinichi0713** [comment]
コードを拝見した限り、**大きな構造的な欠落はありませんが、いくつかの重要な問題点と改善点**が見つかりました。  
特に「学習が進まない」「Q値が全部同じ」という現象につながるポイントを中心に解説します。

---

## 1. **softmaxの使用場所が不適切**
DQN系のQ学習では**出力にsoftmaxをかけてはいけません**。  
Q値は「生（raw）」の値で、softmaxで正規化してしまうと、  
- Q値の“差”が消える  
- 勾配が伝わりにくくなる  
- argmaxが無意味になる  
などの問題が起きます。

**修正：`forward`の最後はsoftmaxを外す**
```python
def forward(self, grid, rects_info):
    grid = grid.to(self.device)
    rects_info = rects_info.to(self.device)
    grid_feat = self.conv(grid)
    rect_feat = self.rect_encoder(rects_info)
    x = torch.cat([grid_feat, rect_feat], dim=1)
    logits = self.fc(x)
    return logits  # ←softmaxは外す
```

---

## 2. **rects_infoの入力サイズミス**
`rects_info`のコメントが `(B, max_rects * 2 + 1)` となっていますが、  
`rects_info` の実際のサイズは `(B, max_rects * 2 + 3)` です（長方形情報 + [num_rects, action_last, reward_last]）。  
**このままではエンコーダの入力次元と合わず、エラーや学習不全の元になります。**

**修正：コメントを合わせる、またはassertでチェックを入れる**
```python
# rects_info: (B, max_rects * 2 + 3)
assert rects_info.shape[1] == self.rect_encoder[0].in_features
```

---

## 3. **ネットワークの重み初期化**
PyTorchのデフォルト初期化で十分ですが、もし全て同じ値が出る場合は  
「重み初期化のバグ」や「.to(self.device)」のタイミングミスも疑ってください。

---

## 4. **バッチノーマライゼーションの未使用**
`self.batch_normalization` をforwardで呼んでいません。  
（コメントアウトされている）

**→ 使わないなら削除してOK。使うなら、`x`に適用する場所を明確に。**

---

## 5. **パラメータロードのタイミング**
`__init__` で `self.__load_state_dict()` を呼んでいますが、  
新規学習時には古いパラメータが勝手にロードされてしまうので注意が必要です。  
（意図的ならOK）

---

## 6. **その他のポイント**
- `self.device` を `super().__init__()` より前に使わないこと
- `.to(self.device)` は`forward`の中で毎回呼ぶより、呼び出し元で揃えた方が効率的

---

## 7. **学習が進まない場合の追加チェック**
- **損失値（loss）が減っているか？**
- **Q値（logits）の分布が広がっているか？**
- **ReplayBufferの内容に多様性があるか？**
- **optimizer.step()が毎回呼ばれているか？**

---

## まとめ（修正版forward例）

```python
def forward(self, grid, rects_info):
    grid = grid.to(self.device)
    rects_info = rects_info.to(self.device)
    grid_feat = self.conv(grid)
    rect_feat = self.rect_encoder(rects_info)
    x = torch.cat([grid_feat, rect_feat], dim=1)
    # x = self.batch_normalization(x)  # 必要なら有効化
    logits = self.fc(x)
    return logits  # Q値そのまま返す
```

---

### **最重要ポイント：Q値にsoftmaxをかけないこと！**
これを修正するだけで、学習・行動選択が大きく改善するはずです。

---

**上記修正後も学習が進まない場合は、損失値やQ値の値をprintしてみてください。  
それでも改善しない場合は、再度詳細を教えてください。**

以上です。
